tests/lib/hardware/openstack_libcloud.py

```diff
@@ -78,19 +79,27 @@ def boot(self):
         )
         logging.debug(f"node {self.name} booted")
 
-        self._create_and_attach_floating_ip()
+        self._get_floating_ip()
         # Wait for node to be ready
         self._wait_until_state(NodeState.RUNNING)
         # Attach a 10GB disk
         self._create_and_attach_volume(10)
 
-    def _create_and_attach_floating_ip(self):
-        # TODO(jhesketh): Move cloud-specific configuration elsewhere
-        floating_ip = self.conn.ex_create_floating_ip(
-            config.OS_EXTERNAL_NETWORK)
-
-        logger.info(f"Created floating IP: {floating_ip}")
-        self._floating_ips.append(floating_ip)
+    def _get_floating_ip(self):
+        try:
+            floating_ip = self.conn.ex_create_floating_ip(
+                config.OS_EXTERNAL_NETWORK)
+            logger.info("Created floating IP: ")
+            logger.info(floating_ip)
+            self._floating_ips.append(floating_ip)
+            self._floating_ips_created.append(floating_ip)
+        except libcloud.common.exceptions.BaseHTTPError:
+            print("Unable to create floating IP")
+            print("Falling back to existing IP association if any is free...")
+            for floating_ip in self.libcloud_conn.ex_list_floating_ips():
+                if floating_ip.node_id is None:
+                    self._floating_ips.append(floating_ip)
```

Comment:
Shouldn't you also assign the floating ip here to the node?

Reply:
It still gets assigned by the lines following because floating_ip is still set.

However, there is no catch for when a floating ip isn't found. In fact, the last one returned will be attempted to be assigned even if it's not available so that needs fixing.

Reply:
Actually I was wondering what is the global expected behavior when some resource allocation fails.
Up until now the behavior I've seen is that the program just continues.
Do you prefer a try/except or a return False ?

Reply:
Raising an exception so that the test fails (and pytest will do the teardown).

I believe all of the other places where we're allocating resources will raise errors on failures which we are deliberately not catching so that they are seen by pytest+the user.